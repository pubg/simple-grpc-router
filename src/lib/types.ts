export type RouterConfig = {
    filters: Filter[],
    nodeId?: string,
    nodeCluster?: string,
    adminPort?: number,
    /** Listener port */
    port?: number,
    /** Listener TLS cert name (sds resource name) */
    certName?: string,
    /** Listener CA cert name (sds resource name) */
    caName?: string,
    /** SDS config resources */
    secrets?: Secret[]
}

export type Filter = {
    path: string,
    host: string,
    port: number,
    certName?: string,
    caName?: string
}

export type Secret = SecretTLS | SecretCA

export type SecretTLS = {
    name: string,
    tls_certificate: {
        certificate_chain: {
            filename: string
        },
        private_key: {
            filename: string
        }
    }
}

export type SecretCA = {
    name: string,
    validation_context: {
        trusted_ca: {
            filename: string
        }
    }
}

export type Route = {
    path: string,
    address: string,
    clusterName: string,
}

export type Cluster = {
    clusterName: string,
    host: string,
    port: number,
    certName?: string,
    caName?: string,
}

export type Config = {
    nodeId: string,
    nodeCluster: string,
    adminPort: number,
    listenPort: number,
    routeList: Route[],
    clusterMap: Map<string, Cluster>,
    outputPath: string,
    sdsConfigResources: Secret[],
    listenerCertName?: string,
    listenerCAName?: string,
}