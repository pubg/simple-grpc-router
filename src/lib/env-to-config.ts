import { Cluster, Config, RouterConfig, Secret } from "./types";

export function envToConfig(rc: RouterConfig): Config {

    const pathSet = new Set<string>();
    const clusterMap = new Map<string, Cluster>();
    const getClusterName = (address: string) => `cluster-${address.replace(/[^a-zA-Z0-9]/g, '-')}`;

    const routeList = rc.filters.map(r => {

        if (pathSet.has(r.path)) throw new Error(`Duplicate path (${r.path})`);
        if(r.caName && !r.certName) throw new Error(`CA name (${r.caName}) without cert name`);
        if(r.caName && !rc.secrets?.find(s => s.name === r.caName)) throw new Error(`CA name (${r.caName}) not found in SDS config`);
        if(r.certName && !rc.secrets?.find(s => s.name === r.certName)) throw new Error(`Cert name (${r.certName}) not found in SDS config`);

        pathSet.add(r.path);
        const address = r.host + ':' + r.port;
        const clusterName = getClusterName(address);

        if (!clusterMap.has(clusterName)) {
            clusterMap.set(clusterName, {
                clusterName,
                host: r.host,
                port: +r.port,
                certName: r.certName,
                caName: r.caName,
            });
        } else {
            const cluster = clusterMap.get(clusterName);
            if (cluster?.certName !== r.certName) throw new Error(`Duplicate cluster name (${clusterName}) with different cert name`);
        }

        return { path: r.path, address, clusterName };
    });

    const sdsConfigResources: Secret[] = rc.secrets || [];

    if(rc.certName && !sdsConfigResources.find(r => r.name === rc.certName))
        throw new Error(`Cert name (${rc.certName}) not found in SDS config`);

    if(rc.caName && !sdsConfigResources.find(r => r.name === rc.caName))
        throw new Error(`CA name (${rc.caName}) not found in SDS config`);

    return {
        nodeId: rc.nodeId || 'envoy',
        nodeCluster: rc.nodeCluster || rc.nodeId || 'envoy',
        adminPort: +(rc.adminPort || 9901),
        listenPort: +(rc.port || 50051),
        clusterMap,
        routeList,
        listenerCertName: rc.certName,
        listenerCAName: rc.caName,
        sdsConfigResources,
        outputPath: '/tmp/',
    };

}
