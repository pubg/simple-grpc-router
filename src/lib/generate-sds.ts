import type { Secret } from "./types";
import type { SDSResources } from "../interface";

export function generateSDS(secret: Secret): SDSResources {
    return {
        resources: [{
            '@type': 'type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret',
            ...secret
        }],
    };
}

export function getSdsCertConfig(name: string) {
    return {
        name,
        sds_config: {
            path: getSdsPath(name),
        }
    }
}

export function getSdsPath(name: string) {
    return `/tmp/sds-${name}.yaml`;
}
