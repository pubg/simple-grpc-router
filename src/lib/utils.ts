import path from 'path';
import dotenv from 'dotenv';
import { debuglog } from 'util';

dotenv.config({ path: path.join(__dirname, '../.env') });

export const INFO = debuglog('INFO');
export const WARN = debuglog('WARN');
export const ERROR = debuglog('ERROR');

process.on('uncaughtException', (err) => {
    console.error(err.message);
    console.error(err.stack);
    process.exit(1);
});