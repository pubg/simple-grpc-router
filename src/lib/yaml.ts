import { dump, load } from 'js-yaml';

export function yamlDump(envoyConfig: any): string {
    return dump(envoyConfig, { noRefs: true });
}

export function yamlParse(yaml: string): any {
    return load(yaml);
}