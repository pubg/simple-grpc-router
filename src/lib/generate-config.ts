import { INFO } from "./utils";
import type { Admin, Cluster, EnvoyConfig, Listener, Node, RouteElement, StaticResources, ListenerTransportSocket, ClusterTransportSocket } from "../interface";
import type { Config } from "./types";
import { getSdsCertConfig } from "./generate-sds";

export function generateConfig(cfg: Config): EnvoyConfig {

    if (!cfg.routeList?.length)
        throw new Error('No routes defined');

    INFO('Config', cfg);

    const node: Node = {
        id: cfg.nodeId,
        cluster: cfg.nodeCluster,
    };

    const admin: Admin = {
        access_log_path: '/tmp/admin_access.log',
        address: {
            socket_address: {
                address: '0.0.0.0',
                port_value: cfg.adminPort,
            },
        }
    };

    const routes: RouteElement[] = cfg.routeList.map(route => ({
        match: { prefix: route.path },
        route: {
            cluster: route.clusterName,
            timeout: '600s',
        },
    }));

    const listenerCert = cfg.listenerCertName ? cfg.sdsConfigResources.find(s => s.name === cfg.listenerCertName) : undefined;
    const listenerCA = cfg.listenerCAName ? cfg.sdsConfigResources.find(s => s.name === cfg.listenerCAName) : undefined;

    if (cfg.listenerCertName && !listenerCert) throw new Error(`Listener cert ${cfg.listenerCertName} not found`);
    if (cfg.listenerCAName && !listenerCA) throw new Error(`Listener CA ${cfg.listenerCAName} not found`);

    const listenerTransportSocket: ListenerTransportSocket | undefined = listenerCert ? {
        name: 'envoy.transport_sockets.tls',
        typed_config: {
            "@type": 'type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext',
            common_tls_context: {
                tls_certificate_sds_secret_configs: getSdsCertConfig(listenerCert.name),
                validation_context_sds_secret_config: cfg.listenerCAName ? getSdsCertConfig(cfg.listenerCAName) : undefined
            }
        }
    } : undefined;

    const listener: Listener = {
        name: 'listener',
        address: {
            socket_address: {
                address: '0.0.0.0',
                port_value: cfg.listenPort,
            },
        },
        filter_chains: [
            {
                transport_socket: listenerTransportSocket,
                filters: [
                    {
                        name: 'envoy.http_connection_manager',
                        typed_config: {
                            "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
                            access_log: [
                                {
                                    name: 'envoy.file_access_log',
                                    typed_config: {
                                        "@type": "type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog",
                                        path: '/dev/stdout',
                                    },
                                }
                            ],
                            stat_prefix: 'grpc_json',
                            codec_type: 'AUTO',
                            route_config: {
                                name: 'local_route',
                                virtual_hosts: [
                                    {
                                        name: 'local_service',
                                        domains: ['*'],
                                        routes: [
                                            ...routes,
                                            {
                                                match: { path: '/' },
                                                direct_response: {
                                                    status: 200,
                                                    body: {
                                                        inline_string: 'ok',
                                                    }
                                                }
                                            },
                                        ],
                                        retry_policy: {
                                            retry_on: 'connect-failure,refused-stream,unavailable',
                                            num_retries: 3,
                                            per_try_timeout: '3600s',
                                            retry_back_off: {
                                                base_interval: '2s',
                                                max_interval: '2s',
                                            },
                                        },
                                    }
                                ],
                            },
                            upgrade_configs: { upgrade_type: 'websocket', enabled: true },
                            http_filters: [{
                                name: 'envoy.filters.http.grpc_web',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.grpc_web.v3.GrpcWeb' }
                            }, {
                                name: 'envoy.filters.http.cors',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.cors.v3.Cors' }
                            }, {
                                name: 'envoy.filters.http.router',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.router.v3.Router' }
                            }]
                        }
                    }
                ],
            },
        ],
    };

    const listeners: Listener[] = [listener];
    const clusters: Cluster[] = [...cfg.clusterMap.values()].map(cluster => {
        const cert = cluster.certName ? cfg.sdsConfigResources.find(r => r.name === cluster.certName) : undefined;
        const ca = cluster.caName ? cfg.sdsConfigResources.find(r => r.name === cluster.caName) : undefined;

        if (cluster.certName && !cert) throw new Error(`Cert ${cluster.certName} not found`);
        if (cluster.caName && !ca) throw new Error(`CA ${cluster.caName} not found`);

        const transport_socket: ClusterTransportSocket | undefined = cert ? {
            name: 'envoy.transport_sockets.tls',
            typed_config: {
                "@type": 'type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext',
                common_tls_context: {
                    tls_certificate_sds_secret_configs: getSdsCertConfig(cert.name),
                    validation_context_sds_secret_config: cluster.caName ? getSdsCertConfig(cluster.caName) : undefined
                },
            },
        } : undefined;

        return {
            name: cluster.clusterName,
            connect_timeout: '1.25s',
            type: 'strict_dns',
            lb_policy: 'round_robin',
            dns_lookup_family: 'V4_ONLY',
            load_assignment: {
                cluster_name: cluster.clusterName,
                endpoints: [{
                    lb_endpoints: [{
                        endpoint: {
                            address: {
                                socket_address: {
                                    address: cluster.host,
                                    port_value: cluster.port,
                                },
                            },
                        }
                    }]
                }]
            },
            http2_protocol_options: {},
            circuit_breakers: {
                thresholds: {
                    max_pending_requests: 100000000,
                    max_requests: 100000000,
                },
            },
            transport_socket,
        }
    });

    const static_resources: StaticResources = {
        listeners,
        clusters,
    }

    return {
        node,
        admin,
        static_resources,
    };
}
