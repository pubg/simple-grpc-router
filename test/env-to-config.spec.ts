import { envToConfig } from '../src/lib/env-to-config';
import type { RouterConfig } from '../src/lib/types';

type TEST = {
    name: string;
    routerConfig: RouterConfig;
}

const defaultEnv = {
};

const tests: TEST[] = [
    {
        name: 'basic',
        routerConfig: {
            ...defaultEnv,
            filters: [
                { path: '/', host: 'localhost', port: 50051 },
                { path: '/test1/', host: 'localhost', port: 50051 },
                { path: '/test2/', host: 'localhost', port: 50053 },
            ],
        }
    },
    {
        name: 'listener-tls',
        routerConfig: {
            ...defaultEnv,
            caName: 'ca-cert',
            certName: 'listener-cert',
            filters: [
                { path: '/', host: 'localhost', port: 50051 },
            ],
            secrets: [
                {
                    name: 'listener-cert',
                    tls_certificate: {
                        certificate_chain: {
                            filename: 'listener-cert/cert',
                        },
                        private_key: {
                            filename: 'listener-cert/key',
                        },
                    },
                },
                {
                    name: 'ca-cert',
                    validation_context: {
                        trusted_ca: {
                            filename: 'ca-cert/ca',
                        },
                    },
                },
            ],
        }
    },
    {
        name: 'cluster-tls',
        routerConfig: {
            ...defaultEnv,
            filters: [
                { path: '/', host: 'localhost', port: 50051, certName: 'cluster-cert', caName: 'ca-cert' },
            ],
            secrets: [
                {
                    name: 'cluster-cert',
                    tls_certificate: {
                        certificate_chain: {
                            filename: 'cluster-cert/cert',
                        },
                        private_key: {
                            filename: 'cluster-cert/key',
                        },
                    },
                },
                {
                    name: 'ca-cert',
                    validation_context: {
                        trusted_ca: {
                            filename: 'ca-cert/ca',
                        },
                    },
                },
            ],
        }
    },
    {
        name: 'mtls',
        routerConfig: {
            ...defaultEnv,
            caName: 'ca-cert',
            certName: 'listener-cert',
            filters: [
                { path: '/', host: 'localhost', port: 50051, certName: 'cluster-cert', caName: 'ca-cert' },
            ],
            secrets: [
                {
                    name: 'cluster-cert',
                    tls_certificate: {
                        certificate_chain: {
                            filename: 'cluster-cert/cert',
                        },
                        private_key: {
                            filename: 'cluster-cert/key',
                        },
                    },
                },
                {
                    name: 'listener-cert',
                    tls_certificate: {
                        certificate_chain: {
                            filename: 'listener-cert/cert',
                        },
                        private_key: {
                            filename: 'listener-cert/key',
                        },
                    },
                },
                {
                    name: 'ca-cert',
                    validation_context: {
                        trusted_ca: {
                            filename: 'ca-cert/ca',
                        },
                    },
                },
            ],
        }
    }
];

describe('envToConfig', () => {
    for (const test of tests) {
        it(test.name, () => {
            expect(envToConfig(test.routerConfig)).toMatchSnapshot();
        });
    }
});
