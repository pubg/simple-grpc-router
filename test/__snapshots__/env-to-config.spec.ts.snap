// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`envToConfig basic 1`] = `
Object {
  "adminPort": 9901,
  "clusterMap": Map {
    "cluster-localhost-50051" => Object {
      "caName": undefined,
      "certName": undefined,
      "clusterName": "cluster-localhost-50051",
      "host": "localhost",
      "port": 50051,
    },
    "cluster-localhost-50053" => Object {
      "caName": undefined,
      "certName": undefined,
      "clusterName": "cluster-localhost-50053",
      "host": "localhost",
      "port": 50053,
    },
  },
  "listenPort": 50051,
  "listenerCAName": undefined,
  "listenerCertName": undefined,
  "nodeCluster": "envoy",
  "nodeId": "envoy",
  "outputPath": "/tmp/",
  "routeList": Array [
    Object {
      "address": "localhost:50051",
      "clusterName": "cluster-localhost-50051",
      "path": "/",
    },
    Object {
      "address": "localhost:50051",
      "clusterName": "cluster-localhost-50051",
      "path": "/test1/",
    },
    Object {
      "address": "localhost:50053",
      "clusterName": "cluster-localhost-50053",
      "path": "/test2/",
    },
  ],
  "sdsConfigResources": Array [],
}
`;

exports[`envToConfig cluster-tls 1`] = `
Object {
  "adminPort": 9901,
  "clusterMap": Map {
    "cluster-localhost-50051" => Object {
      "caName": "ca-cert",
      "certName": "cluster-cert",
      "clusterName": "cluster-localhost-50051",
      "host": "localhost",
      "port": 50051,
    },
  },
  "listenPort": 50051,
  "listenerCAName": undefined,
  "listenerCertName": undefined,
  "nodeCluster": "envoy",
  "nodeId": "envoy",
  "outputPath": "/tmp/",
  "routeList": Array [
    Object {
      "address": "localhost:50051",
      "clusterName": "cluster-localhost-50051",
      "path": "/",
    },
  ],
  "sdsConfigResources": Array [
    Object {
      "name": "cluster-cert",
      "tls_certificate": Object {
        "certificate_chain": Object {
          "filename": "cluster-cert/cert",
        },
        "private_key": Object {
          "filename": "cluster-cert/key",
        },
      },
    },
    Object {
      "name": "ca-cert",
      "validation_context": Object {
        "trusted_ca": Object {
          "filename": "ca-cert/ca",
        },
      },
    },
  ],
}
`;

exports[`envToConfig listener-tls 1`] = `
Object {
  "adminPort": 9901,
  "clusterMap": Map {
    "cluster-localhost-50051" => Object {
      "caName": undefined,
      "certName": undefined,
      "clusterName": "cluster-localhost-50051",
      "host": "localhost",
      "port": 50051,
    },
  },
  "listenPort": 50051,
  "listenerCAName": "ca-cert",
  "listenerCertName": "listener-cert",
  "nodeCluster": "envoy",
  "nodeId": "envoy",
  "outputPath": "/tmp/",
  "routeList": Array [
    Object {
      "address": "localhost:50051",
      "clusterName": "cluster-localhost-50051",
      "path": "/",
    },
  ],
  "sdsConfigResources": Array [
    Object {
      "name": "listener-cert",
      "tls_certificate": Object {
        "certificate_chain": Object {
          "filename": "listener-cert/cert",
        },
        "private_key": Object {
          "filename": "listener-cert/key",
        },
      },
    },
    Object {
      "name": "ca-cert",
      "validation_context": Object {
        "trusted_ca": Object {
          "filename": "ca-cert/ca",
        },
      },
    },
  ],
}
`;

exports[`envToConfig mtls 1`] = `
Object {
  "adminPort": 9901,
  "clusterMap": Map {
    "cluster-localhost-50051" => Object {
      "caName": "ca-cert",
      "certName": "cluster-cert",
      "clusterName": "cluster-localhost-50051",
      "host": "localhost",
      "port": 50051,
    },
  },
  "listenPort": 50051,
  "listenerCAName": "ca-cert",
  "listenerCertName": "listener-cert",
  "nodeCluster": "envoy",
  "nodeId": "envoy",
  "outputPath": "/tmp/",
  "routeList": Array [
    Object {
      "address": "localhost:50051",
      "clusterName": "cluster-localhost-50051",
      "path": "/",
    },
  ],
  "sdsConfigResources": Array [
    Object {
      "name": "cluster-cert",
      "tls_certificate": Object {
        "certificate_chain": Object {
          "filename": "cluster-cert/cert",
        },
        "private_key": Object {
          "filename": "cluster-cert/key",
        },
      },
    },
    Object {
      "name": "listener-cert",
      "tls_certificate": Object {
        "certificate_chain": Object {
          "filename": "listener-cert/cert",
        },
        "private_key": Object {
          "filename": "listener-cert/key",
        },
      },
    },
    Object {
      "name": "ca-cert",
      "validation_context": Object {
        "trusted_ca": Object {
          "filename": "ca-cert/ca",
        },
      },
    },
  ],
}
`;
